src/dotty/tools/dotc/core/Symbols.scala

```diff
@@ -161,6 +161,15 @@ trait Symbols { this: Context =>
           owner.thisType, modcls, parents, decls, TermRef.withSymAndName(owner.thisType, module, name)),
         privateWithin, coord, assocFile)
 
+  def synthesizeCompanionMethod(name: TermName, ret: SymDenotation, owner: SymDenotation)(implicit ctx: Context) = {
+    if(owner.exists && ret.exists) ctx.newSymbol(
+      owner = owner.symbol,
+      name = name,
+      flags = Flags.Synthetic | Flags.Private,
+      info = ExprType(ret.typeRef))
+    else NoSymbol
+  }
+
   /** Create a package symbol with associated package class
```

Comment:
Defining method name based on flags doesn't work here: in order to not force type you will need `flagsUNSAFE`, and tests suggest that Module could still be unset.